[Translation] Added PoEditor Provider

welcoMattic opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | 
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/15310

To follow up on https://github.com/symfony/symfony/pull/38475, this PR adds [PoEditor](https://poeditor.com/) Provider.

The todo list to make it ready is:

- [x] Apply recent changes that have been made on `ProviderInterface` and `TranslatorBagInterface` (we removed the `all()` and `getDomains()` method from TranslatorBagInterface)
- [x] Add PoEditorProvider to `src/Symfony/Bundle/FrameworkBundle/Resources/config/translation_providers.php` file
- [x] Add PoEditor case to `Symfony\Component\Translation\Exception\UnsupportedSchemeException`
- [x] Write integration tests by mocking HTTP Responses

The major part of the remaining work concerns tests, I will make it done before the beginning of May.

welcoMattic committed:
Added PoEditor Provider

stof commented on src/Symfony/Component/Translation/Bridge/PoEditor/Provider/PoEditorProvider.php:
```diff
@@ -0,0 +1,194 @@
+ * In PoEditor:
+ *  * Terms refers to Symfony's translation keys;
+ *  * Translations refers to Symfony's translated messages;
+ *  * Tags refers to Symfony's translation domains
```
`refer` as you use a plural subject in the sentence.

stof commented on src/Symfony/Component/Translation/Bridge/PoEditor/Provider/PoEditorProvider.php:
```diff
@@ -0,0 +1,194 @@
+            }
+        }
+
+        $this->addTranslations($translations[$defaultCatalogue->getLocale()], $defaultCatalogue->getLocale());
```
isn't this writing a single locale and loosing all others ?

welcoMattic replied:
yes, but I already fix this one in the last commit. I will rebase the branch.

welcoMattic commented:
Package / Verify check fails now, it will be fixed as soon as https://github.com/symfony/symfony/pull/40928 will be merged.

OskarStark commented on src/Symfony/Component/Translation/Bridge/PoEditor/PoEditorProvider.php:
```diff
@@ -0,0 +1,204 @@
+ *  * Translations refer to Symfony's translated messages;
+ *  * Tags refer to Symfony's translation domains
+ *
+ * PoEditor's API always returns 200 status code, even in case of failure.
```
🎉 😄

welcoMattic replied:
Fortunately, in all responses body, there is something like 
```
[
    'response' => [
        'status' => 'success',
        'code' => '200',
        'message' => 'OK',
    ],
]
```
:sweat_smile:

OskarStark commented on src/Symfony/Component/Translation/Bridge/PoEditor/PoEditorProviderFactory.php:
```diff
@@ -0,0 +1,66 @@
+ */
+final class PoEditorProviderFactory extends AbstractProviderFactory
+{
+    public const SCHEME = 'poeditor';
```
in mailer and notifier we don't use a const for the scheme

welcoMattic replied:
You fix that in https://github.com/symfony/symfony/pull/40945? I will rebase the PR to fix it on PoEditor Provider and Lokalise Provider. @andrii-bodnar can you also rebase your PR when #40945 will be merged and update Crowdin Provider?

OskarStark replied:
Created a separate PR https://github.com/symfony/symfony/pull/40967

just FYI

OskarStark commented on src/Symfony/Component/Translation/Bridge/PoEditor/composer.json:
```diff
@@ -0,0 +1,30 @@
+{
+    "name": "symfony/po-editor-translation",
```
```suggestion
    "name": "symfony/po-editor-translation-provider",
```

OskarStark commented on src/Symfony/Component/Translation/Exception/UnsupportedSchemeException.php:
```diff
@@ -21,6 +21,10 @@ class UnsupportedSchemeException extends LogicException
         ],
+        'poeditor' => [
+            'class' => Bridge\PoEditor\PoEditorProviderFactory::class,
+            'package' => 'symfony/po-editor-translation',
```
```suggestion
            'package' => 'symfony/po-editor-translation-provider',
```

welcoMattic commented:
It ready for me for a review.
cc @OskarStark @Nyholm @nicolas-grekas

stof commented on src/Symfony/Component/Translation/Bridge/PoEditor/PoEditorProvider.php:
```diff
@@ -0,0 +1,204 @@
+
+        foreach ($locales as $locale) {
+            foreach ($domains as $domain) {
+                $exportResponse = $this->client->request('POST', 'projects/export', [
```
maybe the various request could be initiated first, and then processed as a separate load, to benefit from the built-in concurrency features of symfony/http-client

stof commented on src/Symfony/Component/Translation/Bridge/PoEditor/PoEditorProvider.php:
```diff
@@ -0,0 +1,204 @@
+                    }
+
+                    $deletedIds = $id;
+                    $termsToDelete = [
```
This looks wrong to me. It will delete only the last term

stof commented on src/Symfony/Component/Translation/Bridge/PoEditor/PoEditorProvider.php:
```diff
@@ -0,0 +1,204 @@
+                        continue;
+                    }
+
+                    $deletedIds = $id;
```
should be `$deletedIds[] = $id`

stof commented on src/Symfony/Component/Translation/Bridge/PoEditor/PoEditorProvider.php:
```diff
@@ -0,0 +1,204 @@
+                    ];
+                }
+            }
+            $this->addTranslations($translations[$locale], $locale);
```
This method currently sends an HTTP request and checks the response internally. This prevents benefiting from the built-in concurrency of symfony/http-client for the loop over catalogues.

stof commented on src/Symfony/Component/Translation/Bridge/PoEditor/PoEditorProvider.php:
```diff
@@ -0,0 +1,204 @@
+            $locale = $catalogue->getLocale();
+            foreach ($catalogue->all() as $messages) {
+                foreach ($messages as $id => $message) {
+                    $translations[$locale][] = [
```
This could be simplified by initializing a `$translations` array inside the loop over catalogues and removing the indexing by locale instead.

stof commented on src/Symfony/Component/Translation/Bridge/PoEditor/PoEditorProvider.php:
```diff
@@ -0,0 +1,204 @@
+        foreach ($defaultCatalogue->all() as $domain => $messages) {
+            foreach ($messages as $id => $message) {
+                $terms[] = [
+                    'term' => $id,
```
Be careful. In Symfony, ids are not unique across domains but only inside a domain.

stof commented on src/Symfony/Component/Translation/Test/ProviderTestCase.php:
```diff
@@ -54,11 +55,17 @@ protected function getClient(): MockHttpClient
     }
 
+    /**
+     * @return LoaderInterface|MockObject
```
should be `&` rather than `|`

welcoMattic commented:
@stof I've addressed all your comments. I've added a test for the `delete` method of the Provider (I'll do the same for other providers).
@OskarStark I've rebased the branch on 5.x, and removed the `SCHEME` constant from the Provider Factory.

It looks ready for another review.

EDIT: CI failings seems unrelated

welcoMattic commented:
status: needs review

CI failures seems unrelated to this PR.

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -1342,14 +1343,17 @@ private function registerTranslatorConfiguration(array $config, ContainerBuilder
         foreach ($classToServices as $class => $service) {
-            $package = sprintf('symfony/%s-translation-provider', substr($service, \strlen('translation.provider_factory.')));
+            switch ($package = substr($service, \strlen('translation.provider_factory.'))) {
+                case 'poeditor': $package = 'po-editor'; break;
```
I'd be in favor of naming the package poeditor and skipping this line

welcoMattic replied:
It was the original name, but @fabpot commented about it https://github.com/symfony/symfony/pull/38475#discussion_r588856406
Note that I prefer `poeditor` as package name too, but if we have to follow the same pattern as Notifier and Mailer, it should be `po-editor`

fabpot replied:
My comment was related to the package name. The package name should be directly derived from the bridge name. So, if the bridge name is `PoEditor`, the package name must be `po-editor`.

welcoMattic replied:
I agree, let me know if you prefer to name the bridge `PoEditor` or `Poeditor` (note that the *real* name of the company is POEditor), and I will adapt the package name accordingly.

welcoMattic commented on src/Symfony/Component/Translation/Bridge/PoEditor/Tests/PoEditorProviderTest.php:
```diff
@@ -0,0 +1,514 @@
+     */
+    public function testReadForOneLocaleAndOneDomain(string $locale, string $domain, string $responseContent, TranslatorBag $expectedTranslatorBag)
+    {
+        $loader = $this->getLoader();
+        $loader->expects($this->once())
+            ->method('load')
+            ->willReturn((new XliffFileLoader())->load($responseContent, $locale, $domain));
+
```
I'm not happy with this, but I don't find the correct way to either test it, or mock it. cc @nicolas-grekas  (same in `testReadForManyLocalesAndManyDomains`)

nicolas-grekas replied:
see deps=low failures

fabpot commented:
Tests seem broken.

fabpot commented on src/Symfony/Component/Translation/Bridge/PoEditor/PoEditorProviderFactory.php:
```diff
@@ -0,0 +1,65 @@
+ */
+final class PoEditorProviderFactory extends AbstractProviderFactory
+{
+    private const HOST = 'api.poeditor.com/v2/';
```
Should only be about the host

fabpot commented on src/Symfony/Component/Translation/Bridge/PoEditor/README.md:
```diff
@@ -0,0 +1,28 @@
+Resources
+---------
+
+  * [Contributing](https://symfony.com/doc/current/contributing/index.html)
```
should be one space, not 2

nicolas-grekas replied:
actually, we use two spaces everywhere else... so this is at least consistent

welcoMattic commented on src/Symfony/Component/Translation/CHANGELOG.md:
```diff
@@ -5,6 +5,8 @@ CHANGELOG
 
  * Add `translation:pull` and `translation:push` commands to manage translations with third-party providers
+ * Add `TranslatorBagInterface::getCatalogues` method
+ * Add support to load XLIFF string in `XliffFileLoader`
```
@fabpot I've added those missing changes here

welcoMattic commented:
Last comments have been addressed, branch has been rebased

fabpot commented:
Thank you @welcoMattic.
